fission-web-api/library/Fission/Web/API/User/DID/Types.hs

```diff
@@ -1,18 +1,30 @@
-module Fission.Web.API.User.DID.Types (DID) where
+module Fission.Web.API.User.DID.Types (DID, SetAuthenticated, SetViaChallenge) where
 
+import           Fission.Challenge.Types
 import qualified Fission.Key                as Key
+import           Fission.User.Username
 
 import           Fission.Web.API.Prelude    hiding (Set)
 
 import qualified Fission.Web.API.Auth.Types as Auth
 
-type DID = "did" :> Set
+type DID = "did" :> (SetAuthenticated :<|> SetViaChallenge)
 
-type Set
+type SetAuthenticated
   =  Summary "Update Public Key"
   :> Description "Set currently authenticated user's root public key to another one"
   --
   :> ReqBody '[JSON] Key.Public
   --
   :> Auth.HigherOrder
   :> PutNoContent
+
+type SetViaChallenge
+  =  Summary "Update Public Key via Email Challenge"
+  :> Description "Set given username's public key to something else via providing a challenge obtained from /user/email/recover"
```

Comment:
Optional, but "an idea" — we could use the [`MkLink`](https://www.stackage.org/haddock/lts-18.1/servant-0.18.3/Servant-API.html#t:MkLink) interface to autogenerate the path. It would automatically change if we ever moved that (unlikely), but mainly it'll complain if it breaks.

Reply:
The `SetViaChallenge` route is only used from within the recovery app, not from somewhere in the server.
The server only sends the e-mail and in the email there's a link to the recovery app itself. The app then uses the `SetViaChallenge` route.